src/transformers/data/data_collator.py

```diff
@@ -133,3 +133,111 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
 
         # The rest of the time (10% of the time) we keep the masked input tokens unchanged
         return inputs, labels
+
+
+@dataclass
+class DataCollatorForXLNetLanguageModeling:
+    """
+    Data collator used for language modeling with XLNet.
+    - collates batches of tensors, honoring their tokenizer's pad_token
+    - preprocesses batches for masked language modeling with procedures specific to XLNet
+    """
+
+    tokenizer: PreTrainedTokenizer
+    mlm: bool = True
+    mlm_probability: float = 0.15
+    max_gram: int = 5  # maximum individual mask span length L
+
+    def __call__(self, examples: List[torch.Tensor]) -> Dict[str, torch.Tensor]:
+        batch = self._tensorize_batch(examples)
+        if self.mlm:
+            inputs, perm_mask, target_mapping, labels = self.mask_tokens(batch)
+            return {"input_ids": inputs, "perm_mask": perm_mask, "target_mapping": target_mapping, "labels": labels}
+        else:
+            return {"input_ids": batch, "labels": batch}
```

Comment:
do we really need the `self.mlm` parameter? I would force it to only use `mlm`. Otherwise it's the same as Causal Language Modeling and it would not make sense to use this data collator.

Reply:
Also official it's called `Generalized Autoregressive Pretraining` so it's more CLM than MLM in my opinion. Let's try to only have one option here and give the data collator a good name

Reply:
No, I don't think this is necessary - I was just mimicking how the primary DataCollator works. I'll remove it my next commit and add lines to catch if `self.mlm` is not provided.